## DATA CLEANING CHALLENGE



This is a personal project organized by Promise Nonso for budding Data Ninjas. Although I've learnt the technicalities of data cleaning and wrangling, I haven't been able to work on a personal real world project. I see this challenge as an opportunity. Let's dive in properly!

### About the data set
The datasets provided are two, containing information about the football game players; their individual attributes, their nationatility and the teams they play for, their market value and their wages. The individial attributes are a set of stat variables that shows how good they are in the game.Below are some of the variables;

- Photo_url: link to the picture of the player's picture.
- LongName: Players name. Some are in full and some are short. work on it. 
- Playerurl: Link to players site.
- Nationality: Player's country of origin.
- Position: The wing or position the player has ever played.
- Name: Another column for the name of the player. Notice the some name contains an irregular character. Work on it .
- Age: Players age. 
- OVA: Players overall analysis or rate in percentage
- POT: Players potential in percentage
- Team and Contract
- ID: Players unique Identity number.
- Height: Players height
- Weight: Players weight in lbs
- Foot: Players strong foot
- BOV: Best Overall: Ability of player to hold the comfortably to his feet by percentage rating.
- BP: Ball position: position at which a player possess the ball the most
- Growth: improvement in the game attribute
- Joined: Date at which player started National team
- Value: players worth in Dollars.
- Wage: weekly salary in Dollars 
- Release clause: This is the amount in players contract required for the player to leave to another club, also in dollars.
- FK : Free Kick
- GK: Goal Keeper 
- SM: Skill Moves rating
- A/W: Attacking work rate 
- D/W: Defensive work rate
- W/F: W/F" refers to a player's weak foot rating. It is a measure of how well a player can use their non-dominant foot to - - pass, shoot or dribble. The rating is given on a scale of 1 to 5
- IR: Injury rating also given on a scale of 1 to 5 
- PAS: Pass accuracy given on a 100 rating
- SHO: Shooting Attribute.
- PAC: Pace

Since we're now familiar with the dataset, we can now get our hands dirty and try cleaning our data!

##### Loading the Data

In [111]:
#checking the list of files in the current working directory
import os
wd = os.getcwd()
os.listdir(wd)

['cleaning challenge.ipynb', 'fifa21 raw data v2.csv', 'fifa21_raw_data.csv']

Now we are sure our dataset is in the current worling directory.

In [112]:
#importing pandas library and loading the datasets
import pandas as pd
import numpy as np
data1 = pd.read_csv('fifa21_raw_data.csv')

c:\Users\Moshood Sidiq\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (76) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


### Assessing the Data

We start by going through our data, but visually and programmatically. We are checking for untidiness and messy parts

In [113]:
#checking for the first few rows to get a glimpse of the dataset.
data1

photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/257/710/21_60.png   
18975  https://cdn.sofifa.com/players/258/736/21_60.png   
18976  https://cdn.sofifa.com/players/247/223/21_60.png   
18977  https://cdn.sofifa.com/players/258/760/21_60.png   
18978  https://cdn.sofifa.com/players/255/958/21_60.png   

                           LongName  \
0                      Lionel Messi   
1      C. Ronaldo dos Santos Aveiro   
2                         Jan Oblak   
3                   Kevin De Bruyne   
4        Neymar da Silva Santos Jr.   
...                             ...   
18974                Mengxuan Zhang   
18975                 Vani Da Silva   
18976                        Ao Xia   
18977                     Ben Hough   
18978                  Mateo Flores   

                                               playerUrl Nationality  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   
2      http://sofifa.com/player/200389/jan-oblak/210005/    Slovenia   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   
...                                                  ...         ...   
18974  http://sofifa.com/player/257710/mengxuan-zhang...    China PR   
18975  http://sofifa.com/player/258736/vani-da-silva/...     England   
18976     http://sofifa.com/player/247223/ao-xia/210005/    China PR   
18977  http://sofifa.com/player/258760/ben-hough/210005/     England   
18978  http://sofifa.com/player/255958/mateo-flores/2...     Bolivia   

      Positions               Name  Age  ↓OVA  POT  \
0      RW ST CF           L. Messi   33    93   93   
1         ST LW  Cristiano Ronaldo   35    92   92   
2            GK           J. Oblak   27    91   93   
3        CAM CM       K. De Bruyne   29    91   91   
4        LW CAM          Neymar Jr   28    91   91   
...         ...                ...  ...   ...  ...   
18974        CB     Zhang Mengxuan   21    47   52   
18975        ST        V. Da Silva   17    47   67   
18976        CB             Xia Ao   21    47   55   
18977        CM           B. Hough   17    47   67   
18978       CDM          M. Flores   19    47   63   

                                         Team & Contract  ...     A/W     D/W  \
0                  \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  ...  Medium     Low   
1                      \n\n\n\nJuventus\n2018 ~ 2022\n\n  ...    High     Low   
2               \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  ...  Medium  Medium   
3               \n\n\n\nManchester City\n2015 ~ 2023\n\n  ...    High    High   
4           \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  ...    High  Medium   
...                                                  ...  ...     ...     ...   
18974  \n\n\n\nChongqing Dangdai Lifan FC SWM Team\n2...  ...     Low     Low   
18975           \n\n\n\nOldham Athletic\n2020 ~ 2021\n\n  ...  Medium  Medium   
18976                \n\n\n\nWuhan Zall\n2018 ~ 2022\n\n  ...  Medium  Medium   
18977           \n\n\n\nOldham Athletic\n2020 ~ 2021\n\n  ...  Medium  Medium   
18978              \n\n\n\nClub Bolívar\n2020 ~ 2024\n\n  ...  Medium  Medium   

        IR PAC  SHO PAS  DRI DEF PHY   Hits  
0      5 ★  85   92  91   95  38  65  \n372  
1      5 ★  89   93  81   89  35  77  \n344  
2      3 ★  87   92  78   90  52  90   \n86  
3      4 ★  76   86  93   88  64  78  \n163  
4      5 ★  91   85  86   94  36  59  \n273  
...    ...  ..  ...  ..  ...  ..  ..    ...  
18974  1 ★  58   23  26   27  50  48      2  
18975  1 ★  70

In [114]:
data1[data1['Value']== 105500000]

Empty DataFrame
Columns: [photoUrl, LongName, playerUrl, Nationality, Positions, Name, Age, ↓OVA, POT, Team & Contract, ID, Height, Weight, foot, BOV, BP, Growth, Joined, Loan Date End, Value, Wage, Release Clause, Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power, Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits]
Index: []

[0 rows x 77 columns]

In [115]:
#checking for the summary statistics of numerical variables for possible errors or outliers
description = data1.describe()

description.iloc[:, :20]

Age          ↓OVA           POT             ID           BOV  \
count  18979.000000  18979.000000  18979.000000   18979.000000  18979.000000   
mean      25.194583     65.718636     71.136098  226404.790242     66.751620   
std        4.710753      6.968999      6.114176   27141.673349      6.747017   
min       16.000000     47.000000     47.000000      41.000000     48.000000   
25%       21.000000     61.000000     67.000000  210135.000000     62.000000   
50%       25.000000     66.000000     71.000000  232424.000000     67.000000   
75%       29.000000     70.000000     75.000000  246925.500000     71.000000   
max       53.000000     93.000000     95.000000  259216.000000     93.000000   

             Growth     Attacking      Crossing     Finishing  \
count  18979.000000  18979.000000  18979.000000  18979.000000   
mean       5.417461    248.938616     49.688551     45.842932   
std        5.663954     74.299764     18.131334     19.566978   
min        0.000000     42.000000      6.000000      3.000000   
25%        0.000000    222.000000     38.000000     30.000000   
50%        4.000000    263.000000     54.000000     49.000000   
75%        9.000000    297.000000     63.000000     62.000000   
max       26.000000    437.000000     94.000000     95.000000   

       Heading Accuracy  Short Passing       Volleys         Skill  \
count      18979.000000   18979.000000  18979.000000  18979.000000   
mean          51.941883      58.768007     42.697244    256.481848   
std           17.293626      14.518995     17.647116     78.652897   
min            5.000000       7.000000      3.000000     40.000000   
25%           44.000000      54.000000     30.000000    222.000000   
50%           55.000000      62.000000     44.000000    269.000000   
75%           64.000000      68.000000     56.000000    310.000000   
max           93.000000      94.000000     90.000000    470.000000   

          Dribbling         Curve   FK Accuracy  Long Passing  Ball Control  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      55.587913     47.248116     42.392012     52.725012     58.528795   
std       18.761660     18.208857     17.228069     15.177695     16.566286   
min        5.000000      4.000000      5.000000      5.000000      5.000000   
25%       49.000000     35.000000     31.000000     43.000000     54.000000   
50%       61.000000     49.000000     41.000000     56.000000     63.000000   
75%       68.000000     61.000000     55.000000     64.000000     69.000000   
max       96.000000     94.000000     94.000000     93.000000     96.000000   

           Movement  Acceleration  
count  18979.000000  18979.000000  
mean     317.721007     64.357922  
std       55.881422     14.890208  
min      122.000000     13.000000  
25%      289.000000     57.000000  
50%      327.000000     67.000000  
75%      356.000000     74.000000  
max      464.000000     97.000000

In [116]:
description.iloc[:, 20:41]

Sprint Speed       Agility     Reactions       Balance         Power  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      64.413088     63.367617     61.655619     63.926761    296.606618   
std       14.638874     14.597144      9.072114     14.062703     50.733368   
min       12.000000     14.000000     24.000000     12.000000    122.000000   
25%       57.000000     55.000000     56.000000     56.000000    264.000000   
50%       67.000000     66.000000     62.000000     66.000000    302.000000   
75%       74.000000     74.000000     68.000000     74.000000    334.000000   
max       96.000000     96.000000     95.000000     97.000000    444.000000   

         Shot Power       Jumping       Stamina      Strength    Long Shots  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      57.802413     64.577006     62.654724     64.786132     46.786343   
std       13.308980     11.844592     15.804986     12.488690     19.300862   
min       18.000000     15.000000     12.000000     16.000000      4.000000   
25%       48.000000     58.000000     55.000000     57.000000     32.000000   
50%       59.000000     65.000000     66.000000     66.000000     51.000000   
75%       68.000000     73.000000     73.000000     74.000000     62.000000   
max       95.000000     95.000000     97.000000     97.000000     94.000000   

       ...    Aggression  Interceptions   Positioning        Vision  \
count  ...  18979.000000   18979.000000  18979.000000  18979.000000   
mean   ...     55.590969      46.394067     50.299858     53.857158   
std    ...     17.140038      20.697516     19.429171     13.708607   
min    ...      9.000000       3.000000      2.000000      9.000000   
25%    ...     44.000000      25.000000     40.000000     45.000000   
50%    ...     58.000000      53.000000     55.000000     55.000000   
75%    ...     69.000000      64.000000     64.000000     64.000000   
max    ...     96.000000      91.000000     95.000000     95.000000   

          Penalties     Composure     Defending       Marking  \
count  18979.000000  18979.000000  18979.000000  18979.000000   
mean      48.058328     58.073608    139.908689     46.558828   
std       15.655890     12.079042     61.218624     20.138813   
min        6.000000     12.000000     20.000000      3.000000   
25%       39.000000     50.000000     83.000000     29.000000   
50%       49.000000     59.000000    159.000000     52.000000   
75%       60.000000     67.000000    191.000000     63.000000   
max       92.000000     96.000000    272.000000     94.000000   

       Standing Tackle  Sliding Tackle  
count     18979.000000    18979.000000  
mean         47.705359       45.644502  
std          21.367322       20.922167  
min           5.000000        4.000000  
25%          27.000000       25.000000  
50%          55.000000       52.000000  
75%          65.000000       63.000000  
max          93.000000       90.000000  

[8 rows x 21 columns]

In [117]:
description.iloc[:, 41:]

Goalkeeping     GK Diving   GK Handling    GK Kicking  GK Positioning  \
count  18979.000000  18979.000000  18979.000000  18979.000000    18979.000000   
mean      81.436693     16.416197     16.207124     16.076558       16.217609   
std       84.665579     17.554221     16.816305     16.491064       17.002134   
min       10.000000      2.000000      2.000000      2.000000        2.000000   
25%       48.000000      8.000000      8.000000      8.000000        8.000000   
50%       53.000000     11.000000     11.000000     11.000000       11.000000   
75%       59.000000     14.000000     14.000000     14.000000       14.000000   
max      440.000000     90.000000     92.000000     93.000000       91.000000   

        GK Reflexes   Total Stats    Base Stats           PAC           SHO  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      16.519205   1595.293851    355.703040     67.454239     53.457716   
std       17.854185    269.883542     40.762282     10.678058     13.827229   
min        2.000000    747.000000    232.000000     25.000000     16.000000   
25%        8.000000   1452.000000    327.000000     61.000000     44.000000   
50%       11.000000   1627.000000    356.000000     68.000000     56.000000   
75%       14.000000   1781.000000    384.000000     75.000000     64.000000   
max       90.000000   2316.000000    498.000000     96.000000     93.000000   

                PAS           DRI           DEF           PHY  
count  18979.000000  18979.000000  18979.000000  18979.000000  
mean      57.681069     62.875494     49.865904     64.368618  
std       10.081914      9.927875     16.442730      9.601665  
min       25.000000     25.000000     12.000000     28.000000  
25%       51.000000     57.000000     35.000000     58.000000  
50%       58.000000     64.000000     53.000000     65.000000  
75%       64.000000     69.000000     63.000000     71.000000  
max       93.000000     95.000000     91.000000     91.000000

The discription was splitted so that all the variables can be seen, without some being cut out as there quite alot of them.

In [118]:
#checking for duplicate entries
print(data1.duplicated().any())
print(data1.duplicated().sum())

True
1


In [119]:
data1[data1.duplicated()]

photoUrl       LongName  \
944  https://cdn.sofifa.com/players/251/698/21_60.png  Kevin Berlaso   

                                             playerUrl Nationality Positions  \
944  http://sofifa.com/player/251698/kevin-berlaso/...     Ecuador        RB   

           Name  Age  ↓OVA  POT       Team & Contract  ...   A/W     D/W   IR  \
944  K. Berlaso   32    77   77  \n Ecuador\nFree\n\n  ...  High  Medium  2 ★   

    PAC  SHO PAS  DRI DEF PHY  Hits  
944  78   56  69   77  72  68  \n12  

[1 rows x 77 columns]

In [120]:
data1[data1['LongName']=='Kevin Berlaso']

photoUrl       LongName  \
899  https://cdn.sofifa.com/players/251/698/21_60.png  Kevin Berlaso   
944  https://cdn.sofifa.com/players/251/698/21_60.png  Kevin Berlaso   

                                             playerUrl Nationality Positions  \
899  http://sofifa.com/player/251698/kevin-berlaso/...     Ecuador        RB   
944  http://sofifa.com/player/251698/kevin-berlaso/...     Ecuador        RB   

           Name  Age  ↓OVA  POT       Team & Contract  ...   A/W     D/W   IR  \
899  K. Berlaso   32    77   77  \n Ecuador\nFree\n\n  ...  High  Medium  2 ★   
944  K. Berlaso   32    77   77  \n Ecuador\nFree\n\n  ...  High  Medium  2 ★   

    PAC  SHO PAS  DRI DEF PHY  Hits  
899  78   56  69   77  72  68  \n12  
944  78   56  69   77  72  68  \n12  

[2 rows x 77 columns]

There is only one duplicate from the dataset, Kevin Berlaso.

In [121]:
#checking for null values.
sum(data1.isna().sum())

17966

There are 17,966 null values, most of them probably coming from the Loan Date Column.

In [122]:
data1[['Loan Date End']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Loan Date End  1013 non-null   object
dtypes: object(1)
memory usage: 148.4+ KB


Only 1,013 entries are non-null.

In [123]:
#checking for the info.
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   photoUrl          18979 non-null  object
 1   LongName          18979 non-null  object
 2   playerUrl         18979 non-null  object
 3   Nationality       18979 non-null  object
 4   Positions         18979 non-null  object
 5   Name              18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Team & Contract   18979 non-null  object
 10  ID                18979 non-null  int64 
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  foot              18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  BP                18979 non-null  object
 16  Growth            18979 non-null  int64 
 17  Joined      

#### Documenting the Issues

I was able to figure out about 15 issues I could figure out both in quality and tidiness during assessment. They are as listed below;
- Duplicate entries should be dropped.
- 'Loan Date' column is empty.
- For 'value' and 'Wages' columns, we need to remove the Euro and M/K behind the values.
- 'Photo url' and 'Player url' aren't needed.
- Star signal after values under 'W/F', 'SM', and 'IR'.
- '\n\n\n\n' should be removed from 'club & contract'. The column also needs to be separated.
- '~' should be replaced with '-'.
- height needs to be converted to cm.
- weight also needs to be in kg.
- '\n' in 'Hit' should be removed.
- Positions needs to be separated with commas.
- LongName should be dropped, Name is enough.
- Columns should be better ordered
- The values in the 'Joined' should be converted to international date.
- There are missing values.

#### Cleaning Data

We're done assessing our data. It is now time we get the real work done. It is important we start by making a copy of the dataset should any unforeseen grave mistake happen.

In [124]:
#making a copy
clean_data = data1.copy()

In [125]:
#taking one last look
clean_data.tail(3)

photoUrl      LongName  \
18976  https://cdn.sofifa.com/players/247/223/21_60.png        Ao Xia   
18977  https://cdn.sofifa.com/players/258/760/21_60.png     Ben Hough   
18978  https://cdn.sofifa.com/players/255/958/21_60.png  Mateo Flores   

                                               playerUrl Nationality  \
18976     http://sofifa.com/player/247223/ao-xia/210005/    China PR   
18977  http://sofifa.com/player/258760/ben-hough/210005/     England   
18978  http://sofifa.com/player/255958/mateo-flores/2...     Bolivia   

      Positions       Name  Age  ↓OVA  POT  \
18976        CB     Xia Ao   21    47   55   
18977        CM   B. Hough   17    47   67   
18978       CDM  M. Flores   19    47   63   

                                Team & Contract  ...     A/W     D/W   IR PAC  \
18976       \n\n\n\nWuhan Zall\n2018 ~ 2022\n\n  ...  Medium  Medium  1 ★  64   
18977  \n\n\n\nOldham Athletic\n2020 ~ 2021\n\n  ...  Medium  Medium  1 ★  64   
18978     \n\n\n\nClub Bolívar\n2020 ~ 2024\n\n  ...  Medium  Medium  1 ★  57   

       SHO PAS  DRI DEF PHY Hits  
18976   28  26   38  48  51    3  
18977   40  48   49  35  45    5  
18978   32  43   48  44  49    2  

[3 rows x 77 columns]

We will be the taking the issues one after the other. We are going to be following the pattern;
- Defining the issue.
- Solving it, and
- Testing the solution.

#### - Issue 1
Duplicate entries: we are dropping it.

In [126]:
#dropping duplicate entries
clean_data = clean_data.drop_duplicates()

In [127]:
#testing...
clean_data.duplicated().any()

False

#### - Issue 2

Most of Loan Date column is empty; we can turn the empty column to NaN, and convert the ones that aren't empty to international date standard. I'm thinking of dropping the column completely, but eorking with date of loaned players might be useful when performing analysis.

In [128]:
#taking a look
clean_data['Loan Date End'].value_counts()

Jun 30, 2021    770
Dec 31, 2020    100
May 31, 2021     30
Dec 31, 2021     22
Jan 31, 2021     18
Jan 1, 2021      15
Jun 30, 2022     15
Nov 30, 2020      7
Jun 23, 2021      7
Nov 27, 2021      6
Nov 22, 2020      5
Jan 3, 2021       3
Aug 31, 2021      2
May 31, 2022      2
Jan 4, 2021       2
Jul 1, 2021       1
Dec 30, 2021      1
Jul 5, 2021       1
Jan 17, 2021      1
Jun 30, 2023      1
Jul 31, 2021      1
Dec 30, 2020      1
Jan 30, 2021      1
Aug 1, 2021       1
Name: Loan Date End, dtype: int64

In [129]:
#creating a function...
def change_to_int_date(df, col):
    '''This function changes dates to non-standard version to standard'''
    #creating arrays of month...
    month = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12']
    month2 = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

    #extracting the date...
    month_array = np.array(df[col].str.extract('(^.{0,3})'))

    day_array =  np.array(df[col].str.extract('(\d{1,2})'))
    day_array = np.where(pd.isna(day_array), 'None', day_array)
    day_array = day_array.flatten()

    year_array = np.array(df[col].str.extract('(\d{4})'))
    year_array = np.where(pd.isna(year_array), 'None', year_array)
    year_array = year_array.flatten()

    #converting month_array to number...
    new = []

    for i in month_array:
        if i in month2:
            new.append(month[month2.index(i)])
        else:
            new.append('None')

    new_month = np.array(new)

    #creating new date 
    new_date = year_array + '-' + new_month + '-' + day_array

    #replacing the column with the new values
    df[col] = new_date
    df[col] = df[col].replace('None-None-None', np.NaN, regex=True)
    df[col] = pd.to_datetime(df[col])
    return df[col]

In [130]:
clean_data['Loan Date End'] = change_to_int_date(clean_data, 'Loan Date End')

In [131]:
#testing our code
clean_data['Loan Date End'].value_counts()

2021-06-30    770
2020-12-31    100
2021-05-31     30
2021-12-31     22
2021-01-31     18
2021-01-01     15
2022-06-30     15
2020-11-30      7
2021-06-23      7
2021-11-27      6
2020-11-22      5
2021-01-03      3
2021-08-31      2
2022-05-31      2
2021-01-04      2
2021-07-01      1
2021-12-30      1
2021-07-05      1
2021-01-17      1
2023-06-30      1
2021-07-31      1
2020-12-30      1
2021-01-30      1
2021-08-01      1
Name: Loan Date End, dtype: int64

Our date is now in the standard format.

#### - Issue 3
For 'Value', 'Wages', and 'release clause' column, they need to be only in figures; Euro and M/K should be removed.

In [132]:
#creating a function to reduce repititive codes...
def change_to_num(df, col):
    '''This function changes the shortened values in a column to their proper numerical value'''
    #removing the Euro sign...
    df[col] = df[col].str.replace('€', '').astype(str)
    empty_list = []

    #converting the values to numbers
    for i in df[col]:
        if 'M' in i:
            num = float(i[:-1]) * 1000000
        elif 'K' in i:
            num = float(i[:-1]) * 1000
        else:
            num = float(i)
        empty_list.append(num)
    empty_list = np.array(empty_list)

    #updating our dataframe...
    df[col] = empty_list
    return df[col]

In [133]:
#cleaning the 'Value' column...
clean_data['Value'] = change_to_num(clean_data, 'Value')

In [134]:
#cleaning the 'Wages column...
clean_data['Wage'] = change_to_num(clean_data, 'Wage')

In [135]:
#cleaning 'Release Clause column..
clean_data['Release Clause'] = change_to_num(clean_data, 'Release Clause')

In [136]:
#testing our code...
clean_data[['Value', 'Wage', 'Release Clause']]

Value      Wage  Release Clause
0      67500000.0  560000.0     138400000.0
1      46000000.0  220000.0      75900000.0
2      75000000.0  125000.0     159400000.0
3      87000000.0  370000.0     161000000.0
4      90000000.0  270000.0     166500000.0
...           ...       ...             ...
18974     35000.0    1000.0         57000.0
18975     60000.0     500.0        165000.0
18976     40000.0    1000.0         70000.0
18977     60000.0     500.0        165000.0
18978     60000.0     500.0        167000.0

[18978 rows x 3 columns]

#### - Issue 4
About 3 columns aren't needed; 'LongName' seems redundant to me, 'Player url', 'Photo url'aren't needed too.

In [137]:
#dropping the columns...
clean_data.drop(['LongName', 'playerUrl', 'photoUrl'], axis=1, inplace=True)

In [138]:
#testing our solution...
clean_data.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                                Team & Contract      ID Height  Weight  ...  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  158023   5'7"  159lbs  ...   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   20801   6'2"  183lbs  ...   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  200389   6'2"  192lbs  ...   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  192985  5'11"  154lbs  ...   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  190871   5'9"  150lbs  ...   

      A/W     D/W   IR  PAC SHO PAS  DRI  DEF  PHY   Hits  
0  Medium     Low  5 ★   85  92  91   95   38   65  \n372  
1    High     Low  5 ★   89  93  81   89   35   77  \n344  
2  Medium  Medium  3 ★   87  92  78   90   52   90   \n86  
3    High    High  4 ★   76  86  93   88   64   78  \n163  
4    High  Medium  5 ★   91  85  86   94   36   59  \n273  

[5 rows x 74 columns]

It ws 77 columns before, it is now 74 columns...

#### - Issue 5
Star signal after values under 'W/F', 'SM', and 'IR'. Athough I am still contemplating leaving it as a categorical variable or numerical, probably we resort to changing it to a numerical variable.

In [139]:
#creating a function...
def remove_star(df, col):
    '''This function replaces the star character with an empty string and turns it to int'''
    df[col] = df[col].str.replace('★', '').astype(int)
    return df[col]


In [140]:
#removing star for 'W/F'
clean_data['W/F'] = remove_star(clean_data, 'W/F')

#removing star for 'SM'
clean_data['SM'] = remove_star(clean_data, 'SM')

#removing star in 'IR'
clean_data['IR'] = remove_star(clean_data, 'IR')

In [141]:
#testing our solution...
clean_data[['SM', 'W/F', 'IR']].describe()

SM           W/F            IR
count  18978.000000  18978.000000  18978.000000
mean       2.364949      2.940510      1.091632
std        0.766120      0.668467      0.361769
min        1.000000      1.000000      1.000000
25%        2.000000      3.000000      1.000000
50%        2.000000      3.000000      1.000000
75%        3.000000      3.000000      1.000000
max        5.000000      5.000000      5.000000

The problem is now rectified..

#### -Issue 6
Repeating '\n' in the 'club and contract' column should be removed, it is also necessary to separate the 2 column, having more than 1 info in a column violates tidiness.

In [142]:
#replacing the '\n' with empty string...
clean_data['Team & Contract'] = clean_data['Team & Contract'].str.replace('\n', '').astype(str)

In [143]:
#checking if it is gone
clean_data[['Team & Contract']].sample(5)

Team & Contract
11806            Burton Albion2019 ~ 2021
17407                Cork City2015 ~ 2020
3726   Estudiantes de La Plata2020 ~ 2021
9474           Orlando City SC2019 ~ 2020
13369             Hamburger SV2018 ~ 2021

Now, we can separate the column.

In [144]:
#splitting the column...
clean_data['Team'] = clean_data['Team & Contract'].str.extract(r'(\D+)')
month3 = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
new3 = []

#running a for loop to extract the team names
for i in clean_data['Team']:
    if i[-3:] in month3:
        new3.append(i[0:-3])
    else:
        new3.append(i)

clean_data['Team'] = new3

new4 = []

#running a for loop to extract contacts
for i in clean_data['Team & Contract']:
    if 'Loan' in i:
        new4.append('Loan')
    else:
        new4.append(i[-11:])

clean_data['Contract'] = new4
clean_data['Team'] = clean_data['Team'].str.replace(r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)', '')

#dropping 'Team & Contract' column...
clean_data.drop('Team & Contract', axis=1, inplace=True)

C:\Users\MOSHOO~1\AppData\Local\Temp/ipykernel_9492/2915976748.py:25: FutureWarning: The default value of regex will change from True to False in a future version.
  clean_data['Team'] = clean_data['Team'].str.replace(r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)', '')


To explain what I did above, the team column I extracted above are coming from two kinds of contract; loan contracts and normal contracts. The teams with loan contract still contiains the month their contract is ending, so I had to run a loop to drop the month from the names.As for the loan column, I ran a conditional loop, so it can assign 'Loan' for those that are on loan, and their contract span for those that are not. I ignored the date for loaned contracts, as it is exactly same as in the loan end date column.

In [145]:
#testing the changes...
clean_data[['Team', 'Contract']]

Team     Contract
0                             FC Barcelona  2004 ~ 2021
1                                 Juventus  2018 ~ 2022
2                          Atlético Madrid  2014 ~ 2023
3                          Manchester City  2015 ~ 2023
4                      Paris Saint-Germain  2017 ~ 2022
...                                    ...          ...
18974  Chongqing Dangdai Lifan FC SWM Team  2020 ~ 2020
18975                      Oldham Athletic  2020 ~ 2021
18976                           Wuhan Zall  2018 ~ 2022
18977                      Oldham Athletic  2020 ~ 2021
18978                         Club Bolívar  2020 ~ 2024

[18978 rows x 2 columns]

In [146]:
#testing the changes...
extra = clean_data[~pd.isnull(clean_data['Loan Date End'])][['Team', 'Contract']]


In [147]:
extra

Team Contract
205                Tottenham Hotspur      Loan
250              Paris Saint-Germain      Loan
257                           Fulham      Loan
299                  Atlético Madrid      Loan
305                           Napoli      Loan
...                               ...      ...
18497                   Macarthur FC      Loan
18569  Shijiazhuang Ever Bright F.C.      Loan
18580                      Stevenage      Loan
18638  Shijiazhuang Ever Bright F.C.      Loan
18696               Guangzhou R&F FC      Loan

[1013 rows x 2 columns]

Separating is now complete...

#### - Issue 7
We nedeed to convert the height in feet to cm.

In [148]:
clean_data['Height']

0         5'7"
1         6'2"
2         6'2"
3        5'11"
4         5'9"
         ...  
18974    5'10"
18975     5'7"
18976    5'10"
18977     5'9"
18978     5'9"
Name: Height, Length: 18978, dtype: object

In [149]:
#creating a function that converts feet to cm
def feet_to_cm(x):
    feet, inches = x.split("'")
    total_inches = int(feet)*12 + int(inches[:-1])
    cm = total_inches * 2.54
    return cm
#applying the function to the height column...
clean_data['Height'] = clean_data['Height'].apply(feet_to_cm)

In [150]:
#checking if it works...
clean_data['Height'].describe()

count    18978.000000
mean       181.224543
std          6.818952
min        154.940000
25%        175.260000
50%        180.340000
75%        185.420000
max        205.740000
Name: Height, dtype: float64

Our Height column is now in cm...

#### - Issue 8
The weight column should also be in kg. We are converting the values in lbs to kg, and also make it numerical, rather than string.

In [151]:
#creating a function that convert weight in lbs to kg...
def lbs_to_kg(x):
    return round(float(x.strip('lbs')) / 2.205)

#applying the function to the 'Weight' column.
clean_data['Weight'] = clean_data['Weight'].apply(lbs_to_kg)


In [152]:
#testing our solution...
clean_data['Weight'].describe()

count    18978.000000
mean        75.019022
std          7.073728
min         50.000000
25%         70.000000
50%         75.000000
75%         80.000000
max        110.000000
Name: Weight, dtype: float64

Our weight column looks exactly how we want it to look like now.

#### - Issue 9
'\n' in 'Hit' column should be removed.

In [153]:
#changing all to string...
clean_data['Hits'] = clean_data['Hits'].astype(str)
#replacing the '\n' with an empty string and converting the column to int type
clean_data['Hits'] = clean_data['Hits'].str.extract('(\d+K*)').astype(str)

# I noticed there are values with 'K', it is necessary to convert them to proper numbers too. 
clean_data['Hits'] = change_to_num(clean_data, 'Hits')

In [154]:
#testing...
clean_data['Hits'].describe()

count    18978.000000
mean        15.696965
std         50.955804
min          1.000000
25%          2.000000
50%          3.000000
75%          9.000000
max       2000.000000
Name: Hits, dtype: float64

In [155]:
#Looks like we ahve a kind of outlier here; 4500
clean_data[clean_data['Hits']==4500]

Empty DataFrame
Columns: [Nationality, Positions, Name, Age, ↓OVA, POT, ID, Height, Weight, foot, BOV, BP, Growth, Joined, Loan Date End, Value, Wage, Release Clause, Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power, Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits, Team, Contract]
Index: []

[0 rows x 75 columns]

There are quite a number of high 'Hits' rating I don't really know much about fifa player ratings, so I'll leave it for future questioning. The main problem has been solved.

#### - Issue 10
'Positions' column needs to be separated with commas.

In [156]:
#replacing the white space with comma...
clean_data['Positions'] = clean_data['Positions'].str.replace(' ', ', ')

In [157]:
#testing...
clean_data[['Positions']].sample(5)

Positions
17248        CM
12254        ST
6301    CDM, CM
6497     RM, LM
8780         GK

#### - Issue 11
'Joined' needs to be converted to standard international date time


In [158]:
#talking a look...
clean_data['Joined'].value_counts()

Jul 1, 2019     1344
Jul 1, 2018      865
Jan 1, 2019      682
Jul 1, 2017      520
Jul 1, 2020      435
                ... 
May 17, 2018       1
Feb 25, 2015       1
Jun 1, 2005        1
Sep 10, 2016       1
Feb 28, 2015       1
Name: Joined, Length: 1869, dtype: int64

In [159]:
clean_data['Joined'] = change_to_int_date(clean_data, 'Joined')

In [160]:
#testing...
clean_data['Joined']

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2020-08-01
18975   2020-08-01
18976   2018-07-13
18977   2020-08-01
18978   2020-01-01
Name: Joined, Length: 18978, dtype: datetime64[ns]

We are almost done cooking>>>>

#### - Issue 12

Although we are done dealing with all the quality issues, we are only one step to final tidying. The order of columns should be rearranged. I prefer ID being the first column, with Name and other columns following.

In [161]:
#creating the list of orders...
order =    ['ID','Name','Nationality','Age','↓OVA','POT','Team','Contract','Positions','Height','Weight','foot','BOV',
            'BP','Joined','Loan Date End','Value','Wage','Release Clause','Attacking','Crossing','Finishing','Heading Accuracy',
            'Short Passing','Volleys','Skill','Dribbling','Curve','FK Accuracy','Long Passing','Ball Control','Movement','Acceleration',
            'Sprint Speed','Agility','Reactions','Balance','Power','Shot Power','Jumping','Stamina','Strength','Long Shots','Mentality',
            'Aggression','Interceptions','Positioning','Vision','Penalties','Composure','Defending','Marking','Standing Tackle','Sliding Tackle',
            'Goalkeeping','GK Diving','GK Handling','GK Kicking','GK Positioning','GK Reflexes','Total Stats','Base Stats','W/F','SM','A/W','D/W',
            'IR','PAC','SHO','PAS','DRI','DEF','PHY','Hits']

#creating the new dataframe...
clean_data = clean_data[order]

In [162]:
clean_data.info()  #checking the info

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18978 entries, 0 to 18978
Data columns (total 74 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18978 non-null  int64         
 1   Name              18978 non-null  object        
 2   Nationality       18978 non-null  object        
 3   Age               18978 non-null  int64         
 4   ↓OVA              18978 non-null  int64         
 5   POT               18978 non-null  int64         
 6   Team              18978 non-null  object        
 7   Contract          18978 non-null  object        
 8   Positions         18978 non-null  object        
 9   Height            18978 non-null  float64       
 10  Weight            18978 non-null  int64         
 11  foot              18978 non-null  object        
 12  BOV               18978 non-null  int64         
 13  BP                18978 non-null  object        
 14  Joined            1897

In [163]:
#checking for null values in our cleaned data
clean_data.isna().sum()

ID             0
Name           0
Nationality    0
Age            0
↓OVA           0
              ..
PAS            0
DRI            0
DEF            0
PHY            0
Hits           0
Length: 74, dtype: int64

#### - Issue 13
We are done cleaning the data, but we still need to make some corrections to the column names. They need to be in column name standards.

In [164]:
clean_data.columns = ['ID','Name','Nationality','Age','OVA','POT','Team','Contract','Positions','Height','Weight','foot','BOV',
            'BP','Joined','LoanDateEnd','Value','Wage','ReleaseClause','Attacking','Crossing','Finishing','HeadingAccuracy',
            'ShortPassing','Volleys','Skill','Dribbling','Curve','FK_Accuracy','LongPassing','BallControl','Movement','Acceleration',
            'SprintSpeed','Agility','Reactions','Balance','Power','ShotPower','Jumping','Stamina','Strength','LongShots','Mentality',
            'Aggression','Interceptions','Positioning','Vision','Penalties','Composure','Defending','Marking','StandingTackle','SlidingTackle',
            'Goalkeeping','GK_Diving','GK_Handling','GK_Kicking','GK_Positioning','GK_Reflexes','TotalStats','BaseStats','WF','SM','AW','DW',
            'IR','PAC','SHO','PAS','DRI','DEF','PHY','Hits']


In [165]:
#testing...
clean_data.columns

Index(['ID', 'Name', 'Nationality', 'Age', 'OVA', 'POT', 'Team', 'Contract',
       'Positions', 'Height', 'Weight', 'foot', 'BOV', 'BP', 'Joined',
       'LoanDateEnd', 'Value', 'Wage', 'ReleaseClause', 'Attacking',
       'Crossing', 'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys',
       'Skill', 'Dribbling', 'Curve', 'FK_Accuracy', 'LongPassing',
       'BallControl', 'Movement', 'Acceleration', 'SprintSpeed', 'Agility',
       'Reactions', 'Balance', 'Power', 'ShotPower', 'Jumping', 'Stamina',
       'Strength', 'LongShots', 'Mentality', 'Aggression', 'Interceptions',
       'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending',
       'Marking', 'StandingTackle', 'SlidingTackle', 'Goalkeeping',
       'GK_Diving', 'GK_Handling', 'GK_Kicking', 'GK_Positioning',
       'GK_Reflexes', 'TotalStats', 'BaseStats', 'WF', 'SM', 'AW', 'DW', 'IR',
       'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY', 'Hits'],
      dtype='object')

In [166]:
#taking a one last look...
clean_data

ID               Name Nationality  Age  OVA  POT  \
0      158023           L. Messi   Argentina   33   93   93   
1       20801  Cristiano Ronaldo    Portugal   35   92   92   
2      200389           J. Oblak    Slovenia   27   91   93   
3      192985       K. De Bruyne     Belgium   29   91   91   
4      190871          Neymar Jr      Brazil   28   91   91   
...       ...                ...         ...  ...  ...  ...   
18974  257710     Zhang Mengxuan    China PR   21   47   52   
18975  258736        V. Da Silva     England   17   47   67   
18976  247223             Xia Ao    China PR   21   47   55   
18977  258760           B. Hough     England   17   47   67   
18978  255958          M. Flores     Bolivia   19   47   63   

                                      Team     Contract   Positions  Height  \
0                             FC Barcelona  2004 ~ 2021  RW, ST, CF  170.18   
1                                 Juventus  2018 ~ 2022      ST, LW  187.96   
2                          Atlético Madrid  2014 ~ 2023          GK  187.96   
3                          Manchester City  2015 ~ 2023     CAM, CM  180.34   
4                      Paris Saint-Germain  2017 ~ 2022     LW, CAM  175.26   
...                                    ...          ...         ...     ...   
18974  Chongqing Dangdai Lifan FC SWM Team  2020 ~ 2020          CB  177.80   
18975                      Oldham Athletic  2020 ~ 2021          ST  170.18   
18976                           Wuhan Zall  2018 ~ 2022          CB  177.80   
18977                      Oldham Athletic  2020 ~ 2021          CM  175.26   
18978                         Club Bolívar  2020 ~ 2024         CDM  175.26   

       ...      AW      DW  IR PAC SHO PAS  DRI  DEF  PHY   Hits  
0      ...  Medium     Low   5  85  92  91   95   38   65  372.0  
1      ...    High     Low   5  89  93  81   89   35   77  344.0  
2      ...  Medium  Medium   3  87  92  78   90   52   90   86.0  
3      ...    High    High   4  76  86  93   88   64   78  163.0  
4      ...    High  Medium   5  91  85  86   94   36   59  273.0  
...    ...     ...     ...  ..  ..  ..  ..  ...  ...  ...    ...  
18974  ...     Low     Low   1  58  23  26   27   50   48    2.0  
18975  ...  Medium  Medium   1  70  46  40   53   16   40    3.0  
18976  ...  Medium  Medium   1  64  28  26   38   48   51    3.0  
18977  ...  Medium  Medium   1  64  40  48   49   35   45    5.0  
18978  ...  Medium  Medium   1  57  32  43   48   44   49    2.0  

[18978 rows x 74 columns]

In [167]:
#saving our freshly cooked and cleaned data...

clean_data.to_csv('fifa_clean_data', index=False)

...and this brings us to end of this project, it has being really a wonderful one. I actually learnt quite a bunch from thisssss!!! Big thank you to the organizers for pulling this... Can't wait to get my hands on next data challenges to come.